### ISSUE TITLE
macOS CI builds with OpenMP enabled fail dt-cli self-test

### ISSUE THREAD

**Issue by parafin** [darktable-org/darktable#11934]
See:

- https://github.com/darktable-org/darktable/runs/6644224655?check_suite_focus=true#step:5:43
- https://github.com/darktable-org/darktable/runs/6644224733?check_suite_focus=true#step:5:43
- https://github.com/darktable-org/darktable/runs/6644224811?check_suite_focus=true#step:5:42

It works though with combination of latest macOS and latest Xcode:

- https://github.com/darktable-org/darktable/runs/6644224882?check_suite_focus=true#step:5:43

I've modified CI to ignore the result of this self-test for now in commit https://github.com/darktable-org/darktable/commit/4e3ae961657b75aa2d0574d4d1bc16704b22d2c9 which must be reverted when this issue is fixed. I suspect there's something wrong with libomp package in Homebrew. I don't get crashes with self-build darktable from macports on macOS 10.15 with Xcode 12.4 and deployment target 10.14, which is exactly the first CI configuration (aside from being homebrew-based).

**Comment by jenshannoschwalm** [comment]
Just a note, the logs are not available any more so no chance to analyse ...

**Comment by parafin** [comment]
Just take any CI build.

**Comment by jenshannoschwalm** [comment]
Do you have any detailed logs on a recent build locally?

**Comment by zisoft** [comment]
Seems to be fixed by PR #13727 as well.

Tested on my local fork:

```
        run: |
          ${INSTALL_PREFIX}/bin/darktable --version || true
          ${INSTALL_PREFIX}/bin/darktable-cli \
                 --width 2048 --height 2048 \
                 --hq true --apply-custom-presets false \
                 "${SRC_DIR}/src/tests/integration/images/mire1.cr2" \
                 "${SRC_DIR}/src/tests/integration/0000-nop/nop.xmp" \
                 output.png \
                 --core --disable-opencl --conf host_memory_limit=8192 \
                 --conf worker_threads=4 -t 4 \
                 --conf plugins/lighttable/export/force_lcms2=FALSE \
                 --conf plugins/lighttable/export/iccintent=0

```

```
  compile options:
  bit depth is 64 bit
  normal build
  SSE2 optimized codepath enabled
  OpenMP support enabled
  OpenCL support enabled
  Lua support enabled, API version 9.0.0
  Colord support disabled
  gPhoto2 support enabled
  GraphicsMagick support enabled
  ImageMagick support disabled
  libavif support enabled
  libheif support enabled
  libjxl support enabled
  OpenJPEG support enabled
  OpenEXR support enabled
  WebP support enabled
     0.0001 [dt_init --threads] requested 4 ompthreads restricted to 3
     0.0001 [dt_init --threads] using 3 threads for openmp parallel sections
    14.6968 [export_job] exported to `output.png'

```

**Comment by parafin** [comment]
There seem to be two issues. One of them indeed was fixed #13727. But assertion failure is still present in the logs:
https://github.com/darktable-org/darktable/actions/runs/4341148797/jobs/7580397120#step:5:56
https://github.com/darktable-org/darktable/actions/runs/4341148797/jobs/7580397240#step:5:56
https://github.com/darktable-org/darktable/actions/runs/4341148797/jobs/7580397380#step:5:56

So test case still would fail in CI. And this assertion failure seems to be CI-specific.

**Comment by zisoft** [comment]
strange, this does not happen on the nightly builds:
https://github.com/darktable-org/darktable/actions/runs/4338711568/jobs/7575615261#step:5:48

**Comment by kmilos** [comment]
> strange, this does not happen on the nightly builds

There are no asserts in the Release build?

**Comment by zisoft** [comment]
Ok, rerun the test with debug build type:

`Assertion failed: (!ret), function dt_pthread_mutex_lock_with_caller, file /Users/runner/work/darktable/darktable/src/src/common/dtpthread.h, line 115.`

**Comment by victoryforce** [comment]
@kmilos is of course right, assert does nothing when compiling with NDEBUG (that is, with all builds except Debug).

In our case, assert triggers and signals the failure of the `pthread_mutex_lock()` call. But we don't know exactly what error code this call returned. It would be nice to add a debug print to at least understand what the error was.

It is also interesting why pthread_mutex_lock() returns an error in the macOS build, while there are no problems in the similar Windows build.

**Comment by zisoft** [comment]
so...
I changed the code to get an unconditional log output just prior to the assert:

```
  const double t0 = dt_pthread_get_wtime();
  const int ret = pthread_mutex_lock(&(mutex->mutex));
  
  fprintf(stderr, "[pthread_mutex_lock]: %d, name: %s, file: %s, function: %s\n", ret, mutex->name, file, function);
  assert(!ret);
  
  mutex->time_locked = dt_pthread_get_wtime();

```

Result:

```
[pthread_mutex_lock]: 0, name: /Users/runner/work/darktable/darktable/src/src/control/conf.c:88 (dt_conf_set_if_not_overridden), file: /Users/runner/work/darktable/darktable/src/src/control/conf.c, function: dt_conf_set_if_not_overridden
[pthread_mutex_lock]: 0, name: /Users/runner/work/darktable/darktable/src/src/control/conf.c:88 (dt_conf_set_if_not_overridden), file: /Users/runner/work/darktable/darktable/src/src/control/conf.c, function: dt_conf_set_if_not_overridden
[pthread_mutex_lock]: 0, name: /Users/runner/work/darktable/darktable/src/src/lua/lua.c:162 (dt_lua_unlock_internal), file: /Users/runner/work/darktable/darktable/src/src/lua/lua.c, function: dt_lua_lock_internal
[pthread_mutex_lock]: 0, name: /Users/runner/work/darktable/darktable/src/src/lua/lua.c:149 (dt_lua_lock_internal), file: /Users/runner/work/darktable/darktable/src/src/lua/lua.c, function: dt_lua_unlock_internal
[pthread_mutex_lock]: 0, name: /Users/runner/work/darktable/darktable/src/src/control/control.c:274 (dt_control_running), file: /Users/runner/work/darktable/darktable/src/src/control/control.c, function: dt_control_running
[pthread_mutex_lock]: 22, name: , file: /Users/runner/work/darktable/darktable/src/src/control/control.c, function: dt_control_log
Assertion failed: (!ret), function dt_pthread_mutex_lock_with_caller, file /Users/runner/work/darktable/darktable/src/src/common/dtpthread.h, line 117.

```

Full log can be found here: https://github.com/zisoft/darktable/actions/runs/4351509725/jobs/7603240713#step:5:593

The countless successful mutex locks prior to the failure show that the mutex lock is generally working. 

Error code 22 is
22 | EINVAL | Invalid argument

**Comment by parafin** [comment]
Name is missing, because this mutex is uninitialized. It's initialized in function dt_control_init, which doesn't seem to be called for dt-cli:
https://github.com/darktable-org/darktable/blob/ede51437d4cb14132177003b47bc78d273e79580/src/common/darktable.c#L1209-L1212

So this is a logic error, not specific to macOS.

**Comment by zisoft** [comment]
...which leads to the question why this does not occur on linux or windows

**Comment by zisoft** [comment]
Just started a test, I will report here.

BTW: Linux CI is built with target Release...
https://github.com/darktable-org/darktable/blob/5f62d310ea3aac1c408337aa10b548464817d9f8/.github/workflows/ci.yml#L56-L57

**Comment by kmilos** [comment]
> BTW: Linux CI is built with target Release...

And `winpthreads` implementation is definitely different (i.e. unique) compared to Linux and macOS...

**Comment by parafin** [comment]
macOS pthread implementation differs to Linux one too. macOS is closer to BSD.

**Comment by zisoft** [comment]
Linux gives a successful return code 0 with that unitialized mutex:
https://github.com/zisoft/darktable/actions/runs/4352436229/jobs/7605222123#step:6:575

Windows as well, return code 0:
https://github.com/zisoft/darktable/actions/runs/4352436229/jobs/7605220970#step:6:607

So it seems that only the macOS implementation handles that correctly.

**Comment by parafin** [comment]
I wouldn't say Linux/Windows handles it incorrectly, behaviour is undefined for uninitialized mutex, meaning anything can happen. Furthermore there is static initializer for mutex - PTHREAD_MUTEX_INITIALIZER, which may match with just zeroed-out memory. It's just not the case for macOS.

**Comment by zisoft** [comment]
Anyway, an `assert()` on a return value which may or may not be undefined does not give any benefit

**Comment by parafin** [comment]
Does it not? We found a bug in dt code thanks to it.

**Comment by zisoft** [comment]
That's correct and now we can hopefully stop blaming macOS for the failing CI 🙂

**Comment by victoryforce** [comment]
Originally this issue was about OpenMP-related crash fixed by limiting the number of ompthreads to possilbe value. This is fixed in #13736.

The problem investigated in the comments was about mutex-related failed assertion. This is fixed in #13852 by adding the missing mutex init.

As both problems mentioned now fixed I suppose we can close this issue?
